source/MaterialXGenShader/HwLightHandler.cpp

```diff
@@ -33,6 +28,7 @@ HwLightHandler::~HwLightHandler()
 void HwLightHandler::addLightSource(NodePtr node)
 {
     _lightSources.push_back(node);
+    _lightCategories.insert(node->getCategory());
 }
 
 void HwLightHandler::bindLightShaders(HwShaderGenerator& shadergen, const GenOptions& options) const
```

Comment:
The loop below will try to bind a light shader for each light instance. So for multiple instances of the same type it will recreate a new ShaderNodeImplPtr for each. It would be better if this loop was over _lightCategories instead. Maybe we should save pointers to NodeDefs in the set instead of category strings?

Reply:
I think the easiest thing to do is optimize bindLightShader(). It does not check to see if the type has been added or not yet and wastes time retrying to find implementations to override the same _boundLightShaders[id] slot. The actual emit code seems fine since it uses this list -- so it seems it's the filling of this list which is inefficient. Is this a correct understanding?

Reply:
Yes exactly, it works now but it's inefficient to recreate them multiple times. I agree we should refactor bindLightShaders, we've discussed that before and this would be a good time. Let's sync offline.